### ISSUE TITLE
Building on Linux fails

### ISSUE THREAD

**Issue by Frando** [Schachigel/DKV2#6]
Hi,

thanks for this tool. I'm watching your presentation :slightly_smiling_face:  and now tried building the project on Linux.

As there are no build instructions, and I'm not experienced with C++/Qt setups, I might be missing something obvious.

Here's what I did:

- Install `qtcreator`
- Open `dkv2.pro` with qtcreator
- Click the *Run* button

`qmake` seems to run fine without errors, but the actual build then errors with the following log.

Environment:
* Arch linux
* QT Creator 4.14.1
* main branch as of 2021-02-25

```
21:13:29: Running steps for project dkv2...
21:13:29: Configuration unchanged, skipping qmake step.
21:13:29: Starting: "/usr/bin/make" -j8
cd DKV2/ && ( test -e Makefile || /usr/bin/qmake -o Makefile /home/bit/Code/mhs/DKV2/DKV2/DKV2.pro -spec linux-g++ CONFIG+=debug CONFIG+=qml_debug ) && /usr/bin/make -f Makefile 
cd TESTS/ && ( test -e Makefile || /usr/bin/qmake -o Makefile /home/bit/Code/mhs/DKV2/TESTS/TESTS.pro -spec linux-g++ CONFIG+=debug CONFIG+=qml_debug ) && /usr/bin/make -f Makefile 
make[1]: Entering directory '/home/bit/Code/mhs/build-dkv2-Desktop-Debug/TESTS'
g++ -c -pipe -g -std=gnu++1y -Wall -Wextra -D_REENTRANT -fPIC -DQT_DEPRECATED_WARNINGS -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_TESTLIB_LIB -DQT_SQL_LIB -DQT_CORE_LIB -DQT_TESTCASE_BUILDDIR='"/home/bit/Code/mhs/build-dkv2-Desktop-Debug/TESTS"' -DQT_QML_DEBUG -I../../DKV2/TESTS -I. -I/usr/include/qt -I/usr/include/qt/QtPrintSupport -I/usr/include/qt/QtWidgets -I/usr/include/qt/QtGui -I/usr/include/qt/QtTest -I/usr/include/qt/QtSql -I/usr/include/qt/QtCore -I. -I/usr/lib/qt/mkspecs/linux-g++ -o booking.o ../../DKV2/DKV2/booking.cpp
g++ -c -pipe -g -std=gnu++1y -Wall -Wextra -D_REENTRANT -fPIC -DQT_DEPRECATED_WARNINGS -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_TESTLIB_LIB -DQT_SQL_LIB -DQT_CORE_LIB -DQT_TESTCASE_BUILDDIR='"/home/bit/Code/mhs/build-dkv2-Desktop-Debug/TESTS"' -DQT_QML_DEBUG -I../../DKV2/TESTS -I. -I/usr/include/qt -I/usr/include/qt/QtPrintSupport -I/usr/include/qt/QtWidgets -I/usr/include/qt/QtGui -I/usr/include/qt/QtTest -I/usr/include/qt/QtSql -I/usr/include/qt/QtCore -I. -I/usr/lib/qt/mkspecs/linux-g++ -o contract.o ../../DKV2/DKV2/contract.cpp
g++ -c -pipe -g -std=gnu++1y -Wall -Wextra -D_REENTRANT -fPIC -DQT_DEPRECATED_WARNINGS -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_TESTLIB_LIB -DQT_SQL_LIB -DQT_CORE_LIB -DQT_TESTCASE_BUILDDIR='"/home/bit/Code/mhs/build-dkv2-Desktop-Debug/TESTS"' -DQT_QML_DEBUG -I../../DKV2/TESTS -I. -I/usr/include/qt -I/usr/include/qt/QtPrintSupport -I/usr/include/qt/QtWidgets -I/usr/include/qt/QtGui -I/usr/include/qt/QtTest -I/usr/include/qt/QtSql -I/usr/include/qt/QtCore -I. -I/usr/lib/qt/mkspecs/linux-g++ -o dbstatistics.o ../../DKV2/DKV2/dbstatistics.cpp
g++ -c -pipe -g -std=gnu++1y -Wall -Wextra -D_REENTRANT -fPIC -DQT_DEPRECATED_WARNINGS -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_TESTLIB_LIB -DQT_SQL_LIB -DQT_CORE_LIB -DQT_TESTCASE_BUILDDIR='"/home/bit/Code/mhs/build-dkv2-Desktop-Debug/TESTS"' -DQT_QML_DEBUG -I../../DKV2/TESTS -I. -I/usr/include/qt -I/usr/include/qt/QtPrintSupport -I/usr/include/qt/QtWidgets -I/usr/include/qt/QtGui -I/usr/include/qt/QtTest -I/usr/include/qt/QtSql -I/usr/include/qt/QtCore -I. -I/usr/lib/qt/mkspecs/linux-g++ -o dkdbhelper.o ../../DKV2/DKV2/dkdbhelper.cpp
g++ -c -pipe -g -std=gnu++1y -Wall -Wextra -D_REENTRANT -fPIC -DQT_DEPRECATED_WARNINGS -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_TESTLIB_LIB -DQT_SQL_LIB -DQT_CORE_LIB -DQT_TESTCASE_BUILDDIR='"/home/bit/Code/mhs/build-dkv2-Desktop-Debug/TESTS"' -DQT_QML_DEBUG -I../../DKV2/TESTS -I. -I/usr/include/qt -I/usr/include/qt/QtPrintSupport -I/usr/include/qt/QtWidgets -I/usr/include/qt/QtGui -I/usr/include/qt/QtTest -I/usr/include/qt/QtSql -I/usr/include/qt/QtCore -I. -I/usr/lib/qt/mkspecs/linux-g++ -o dkdbviews.o ../../DKV2/DKV2/dkdbviews.cpp
g++ -c -pipe -g -std=gnu++1y -Wall -Wextra -D_REENTRANT -fPIC -DQT_DEPRECATED_WARNINGS -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_TESTLIB_LIB -DQT_SQL_LIB -DQT_CORE_LIB -DQT_TESTCASE_BUILDDIR='"/home/bit/Code/mhs/build-dkv2-Desktop-Debug/TESTS"' -DQT_QML_DEBUG -I../../DKV2/TESTS -I. -I/usr/include/qt -I/usr/include/qt/QtPrintSupport -I/usr/include/qt/QtWidgets -I/usr/include/qt/QtGui -I/usr/include/qt/QtTest -I/usr/include/qt/QtSql -I/usr/include/qt/QtCore -I. -I/usr/lib/qt/mkspecs/linux-g++ -o helperfin.o ../../DKV2/DKV2/helperfin.cpp
g++ -c -pipe -g -std=gnu++1y -Wall -Wextra -D_REENTRANT -fPIC -DQT_DEPRECATED_WARNINGS -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_TESTLIB_LIB -DQT_SQL_LIB -DQT_CORE_LIB -DQT_TESTCASE_BUILDDIR='"/home/bit/Code/mhs/build-dkv2-Desktop-Debug/TESTS"' -DQT_QML_DEBUG -I../../DKV2/TESTS -I. -I/usr/include/qt -I/usr/include/qt/QtPrintSupport -I/usr/include/qt/QtWidgets -I/usr/include/qt/QtGui -I/usr/include/qt/QtTest -I/usr/include/qt/QtSql -I/usr/include/qt/QtCore -I. -I/usr/lib/qt/mkspecs/linux-g++ -o letterTemplate.o ../../DKV2/DKV2/letterTemplate.cpp
make[1]: Entering directory '/home/bit/Code/mhs/build-dkv2-Desktop-Debug/DKV2'
g++ -c -pipe -g -std=gnu++1y -Wall -Wextra -D_REENTRANT -fPIC -DQT_DEPRECATED_WARNINGS -DQT_USE_QSTRINGBUILDER -DQT_QML_DEBUG -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_SQL_LIB -DQT_CORE_LIB -I../../DKV2/DKV2 -I. -I/usr/include/qt -I/usr/include/qt/QtPrintSupport -I/usr/include/qt/QtWidgets -I/usr/include/qt/QtGui -I/usr/include/qt/QtSql -I/usr/include/qt/QtCore -I. -I. -I/usr/lib/qt/mkspecs/linux-g++ -o booking.o ../../DKV2/DKV2/booking.cpp
In file included from ../../DKV2/DKV2/booking.cpp:4:
../../DKV2/DKV2/contract.h:79:19: error: declaration of ‘interestModel contract::interestModel() const’ changes meaning of ‘interestModel’ [-fpermissive]
   79 |     interestModel interestModel() const { return fromInt(td.getValue(qsl("thesaurierend")).toInt());}
      |                   ^~~~~~~~~~~~~
../../DKV2/DKV2/contract.h:15:12: note: ‘interestModel’ declared here as ‘enum class interestModel’
   15 | enum class interestModel {
      |            ^~~~~~~~~~~~~
In file included from ../../DKV2/DKV2/dbstatistics.h:8,
                 from ../../DKV2/DKV2/dbstatistics.cpp:5:
../../DKV2/DKV2/contract.h:79:19: error: declaration of ‘interestModel contract::interestModel() const’ changes meaning of ‘interestModel’ [-fpermissive]
   79 |     interestModel interestModel() const { return fromInt(td.getValue(qsl("thesaurierend")).toInt());}
      |                   ^~~~~~~~~~~~~
../../DKV2/DKV2/contract.h:15:12: note: ‘interestModel’ declared here as ‘enum class interestModel’
   15 | enum class interestModel {
      |            ^~~~~~~~~~~~~
In file included from ../../DKV2/DKV2/booking.cpp:4:
../../DKV2/DKV2/contract.h:79:19: error: declaration of ‘interestModel contract::interestModel() const’ changes meaning of ‘interestModel’ [-fpermissive]
   79 |     interestModel interestModel() const { return fromInt(td.getValue(qsl("thesaurierend")).toInt());}
      |                   ^~~~~~~~~~~~~
../../DKV2/DKV2/contract.h:15:12: note: ‘interestModel’ declared here as ‘enum class interestModel’
   15 | enum class interestModel {
      |            ^~~~~~~~~~~~~
make[1]: Leaving directory '/home/bit/Code/mhs/build-dkv2-Desktop-Debug/DKV2'
make[1]: *** [Makefile:956: booking.o] Error 1
make: *** [Makefile:48: sub-DKV2-make_first] Error 2
make: *** Waiting for unfinished jobs....
g++ -c -pipe -g -std=gnu++1y -Wall -Wextra -D_REENTRANT -fPIC -DQT_DEPRECATED_WARNINGS -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_TESTLIB_LIB -DQT_SQL_LIB -DQT_CORE_LIB -DQT_TESTCASE_BUILDDIR='"/home/bit/Code/mhs/build-dkv2-Desktop-Debug/TESTS"' -DQT_QML_DEBUG -I../../DKV2/TESTS -I. -I/usr/include/qt -I/usr/include/qt/QtPrintSupport -I/usr/include/qt/QtWidgets -I/usr/include/qt/QtGui -I/usr/include/qt/QtTest -I/usr/include/qt/QtSql -I/usr/include/qt/QtCore -I. -I/usr/lib/qt/mkspecs/linux-g++ -o helpersql.o ../../DKV2/DKV2/helpersql.cpp
In file included from ../../DKV2/DKV2/contract.cpp:7:
../../DKV2/DKV2/contract.h:79:19: error: declaration of ‘interestModel contract::interestModel() const’ changes meaning of ‘interestModel’ [-fpermissive]
   79 |     interestModel interestModel() const { return fromInt(td.getValue(qsl("thesaurierend")).toInt());}
      |                   ^~~~~~~~~~~~~
../../DKV2/DKV2/contract.h:15:12: note: ‘interestModel’ declared here as ‘enum class interestModel’
   15 | enum class interestModel {
      |            ^~~~~~~~~~~~~
In file included from ../../DKV2/DKV2/dkdbhelper.cpp:13:
../../DKV2/DKV2/contract.h:79:19: error: declaration of ‘interestModel contract::interestModel() const’ changes meaning of ‘interestModel’ [-fpermissive]
   79 |     interestModel interestModel() const { return fromInt(td.getValue(qsl("thesaurierend")).toInt());}
      |                   ^~~~~~~~~~~~~
../../DKV2/DKV2/contract.h:15:12: note: ‘interestModel’ declared here as ‘enum class interestModel’
   15 | enum class interestModel {
      |            ^~~~~~~~~~~~~
g++ -c -pipe -g -std=gnu++1y -Wall -Wextra -D_REENTRANT -fPIC -DQT_DEPRECATED_WARNINGS -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_TESTLIB_LIB -DQT_SQL_LIB -DQT_CORE_LIB -DQT_TESTCASE_BUILDDIR='"/home/bit/Code/mhs/build-dkv2-Desktop-Debug/TESTS"' -DQT_QML_DEBUG -I../../DKV2/TESTS -I. -I/usr/include/qt -I/usr/include/qt/QtPrintSupport -I/usr/include/qt/QtWidgets -I/usr/include/qt/QtGui -I/usr/include/qt/QtTest -I/usr/include/qt/QtSql -I/usr/include/qt/QtCore -I. -I/usr/lib/qt/mkspecs/linux-g++ -o csvwriter.o ../../DKV2/DKV2/csvwriter.cpp
make[1]: *** [Makefile:998: dbstatistics.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:956: booking.o] Error 1
make[1]: *** [Makefile:970: contract.o] Error 1
make[1]: *** [Makefile:1029: dkdbhelper.o] Error 1
make[1]: Leaving directory '/home/bit/Code/mhs/build-dkv2-Desktop-Debug/TESTS'
make: *** [Makefile:73: sub-TESTS-make_first] Error 2
21:13:38: The process "/usr/bin/make" exited with code 2.
Error while building/deploying project dkv2 (kit: Desktop)
When executing step "Make"
21:13:38: Elapsed time: 00:09.
```

**Comment by Frando** [comment]
When invoking `make` directly in the build dir created by `qmake`, the build log is shorter, but refers to the same error likely:

```
cd DKV2/ && ( test -e Makefile || /usr/bin/qmake -o Makefile /home/bit/Code/mhs/DKV2/DKV2/DKV2.pro -spec linux-g++ CONFIG+=debug CONFIG+=qml_debug ) && make -f Makefile
make[1]: Verzeichnis „/home/bit/Code/mhs/build-dkv2-Desktop-Debug/DKV2“ wird betreten
g++ -c -pipe -g -std=gnu++1y -Wall -Wextra -D_REENTRANT -fPIC -DQT_DEPRECATED_WARNINGS -DQT_USE_QSTRINGBUILDER -DQT_QML_DEBUG -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_SQL_LIB -DQT_CORE_LIB -I../../DKV2/DKV2 -I. -I/usr/include/qt -I/usr/include/qt/QtPrintSupport -I/usr/include/qt/QtWidgets -I/usr/include/qt/QtGui -I/usr/include/qt/QtSql -I/usr/include/qt/QtCore -I. -I. -I/usr/lib/qt/mkspecs/linux-g++ -o booking.o ../../DKV2/DKV2/booking.cpp
In Datei, eingebunden von ../../DKV2/DKV2/booking.cpp:4:
../../DKV2/DKV2/contract.h:79:19: Fehler: Deklaration von »interestModel contract::interestModel() const« ändert die Bedeutung von »interestModel« [-fpermissive]
   79 |     interestModel interestModel() const { return fromInt(td.getValue(qsl("thesaurierend")).toInt());}
      |                   ^~~~~~~~~~~~~
../../DKV2/DKV2/contract.h:15:12: Anmerkung: »interestModel« wurde hier als »enum class interestModel« deklariert
   15 | enum class interestModel {
      |            ^~~~~~~~~~~~~
make[1]: *** [Makefile:956: booking.o] Fehler 1
make[1]: Verzeichnis „/home/bit/Code/mhs/build-dkv2-Desktop-Debug/DKV2“ wird verlassen
make: *** [Makefile:48: sub-DKV2-make_first] Fehler 2

```

**Comment by Frando** [comment]
OK, so I looked for the commit that introduced `interestModel` which seems to be 226d9f04714cac88e91613fdf91d35136bab256a. When reverting to the commit before that, 087564b92c730d17ce40591ce545f79f2ba66bde, the build suceeds and the resulting binary work fine.

**Comment by Schachigel** [comment]
Working with older versions is usually not a good idea ;)

I am not familiar with the build on linux - however the build log you postet suggests, that the compiler option "-fpermissive" might solve your problem. 

Or you can try the latest commit I just released

**Comment by Frando** [comment]
Nice, your latest commit fixed it, the g++ build succeeds.

I got another error from the linker that it couldn't find `lVersion`. I installed `libversion`, but that apparently didn't help. Maybe it's a misconfiguration on my machine, I'll try to check. While trying to find out what's going wrong I commented-out the `LIBS += -lVersion` statement in the `DKV2.pro` build file. That build succeeded and the binary runs fine, I could create a database and everything seems to work great so far. :slightly_smiling_face: 
Maybe `lVersion` is not needed on Linux? I'll try to find out also why the `libversion` is not found.

I could also look into adding some build instructions, and/or setting up an automated build, e.g. via Github actions?
